object.go

```diff
@@ -0,0 +1,706 @@
+	Delete(name string) error
+
+	// AddLink will add a link to another object into this object store.
+	AddLink(name string, obj *ObjectInfo) (*ObjectInfo, error)
+
+	// AddBucket will add a link to another object store.
+	AddBucket(name string, bucket ObjectStore) (*ObjectInfo, error)
+
+	// Seal will seal the object store, no further modifications will be allowed.
+	Seal() error
+
+	// Watch for changes in the underlying store and receive meta information updates.
+	Watch(cb ObjectStoreUpdate) (*Subscription, error)
+}
+
+type ObjectStoreUpdate func(meta *ObjectInfo)
+
+var (
+	errBadMeta        = errors.New("nats: object stream meta information invalid")
+	errBadObjectName  = errors.New("nats: invalid object name")
+	errDigestMismatch = errors.New("nats: received corrupt object, digests do not match")
+)
+
+// ObjectStoreConfig is the config for the object store.
+type ObjectStoreConfig struct {
+	Bucket      string
+	Description string
+	TTL         time.Duration
+	Storage     StorageType
+	Replicas    int
+}
+
+// ObjectMetaOptions
+type ObjectMetaOptions struct {
+	Link      *ObjectLink `json:"link,omitempty"`
+	ChunkSize uint32      `json:"max_chunk_size,omitempty"`
+}
+
+// ObjectMeta is high level information about an object.
+type ObjectMeta struct {
+	Name        string `json:"name"`
+	Description string `json:"description,omitempty"`
+	Headers     Header `json:"headers,omitempty"`
+
+	// Optional options.
+	Opts *ObjectMetaOptions `json:"options,omitempty"`
+}
+
+// ObjectInfo is meta plus instance information.
+type ObjectInfo struct {
+	ObjectMeta
+	Bucket  string    `json:"bucket"`
+	NUID    string    `json:"nuid"`
+	Size    uint64    `json:"size"`
+	ModTime time.Time `json:"mtime"`
+	Chunks  uint32    `json:"chunks"`
+	Digest  string    `json:"digest,omitempty"`
+	Deleted bool      `json:"deleted,omitempty"`
+}
+
+// ObjectLink is used to embed links to other buckets and objects.
+type ObjectLink struct {
+	// Bucket is the name of the other object store.
+	Bucket string `json:"bucket"`
+	// Name can be used to link to a single object.
+	// If empty means this is a link to the whole store, like a directory.
+	Name string `json:"name,omitempty"`
+}
+
+// ObjectResult will return the underlying stream info and also be an io.ReadCloser.
+type ObjectResult interface {
+	io.ReadCloser
+	Info() (*ObjectInfo, error)
+	Error() error
+}
+
+const (
+	objNameTmpl         = "OBJ_%s"
+	objSubjectsPre      = "$O."
+	objAllChunksPreTmpl = "$O.%s.C.>"
+	objAllMetaPreTmpl   = "$O.%s.M.>"
+	objChunksPreTmpl    = "$O.%s.C.%s"
+	objMetaPreTmpl      = "$O.%s.M.%s"
+	objNoPending        = "0"
+	objDefaultChunkSize = uint32(128 * 1024) // 128k
+	objDigestType       = "sha-256="
+	objDigestTmpl       = objDigestType + "%s"
+)
+
+type obs struct {
+	name   string
+	stream string
+	js     *js
+}
+
+// CreateObjectStore will create an object store.
+func (js *js) CreateObjectStore(cfg *ObjectStoreConfig) (ObjectStore, error) {
```

Comment:
I think maybe we should update the interface to have the partial opts like:
```suggestion
func (js *js) CreateObjectStore(cfg *ObjectStoreConfig, opts ...JSOpt) (ObjectStore, error) {
```
so that later on can pass the arguments below for example like: `js.AddStream(scfg, opts)`
That would make it possible to change the timeout or context for example.